Workflow-to-Job Scheduling Parameters

JeffreyThiessen opened the pull request:
This is the first attempt at a solution for galaxyproject#4637.
The input parameters the user can enter are being populated by the value/label pairs being defined in workflows.py but this will need to be changed to use xml parsing like job parameters use.

@Takadonet @jmchilton

JeffreyThiessen committed:
Workflow-to-Job Scheduling Parameters

JeffreyThiessen committed:
PEP8 fixes

JeffreyThiessen committed:
Update DTD test object

JeffreyThiessen committed:
xml parsing

JeffreyThiessen committed:
update dev branch

JeffreyThiessen committed:
update from upstream

JeffreyThiessen committed:
update static files

JeffreyThiessen commented on lib/galaxy/managers/workflows.py:
```diff
@@ -420,14 +421,143 @@ def _workflow_to_dict_run(self, trans, stored):
+
+        return params
+
+    def _workflow_options_xml_validation(self):
```
Would this function be better suited in another file?
Also, I would like to allow this to be callable by an admin to verify if their file well formed.

JeffreyThiessen committed:
pep8

JeffreyThiessen committed:
small fix and refactor

JeffreyThiessen committed:
Validate users allowed priorities before running job from DTD, + Refactoring

JeffreyThiessen committed:
fix sample file

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -4176,16 +4176,17 @@ def to_dict(self, view='collection', value_mapper=None):
+    dict_collection_visible_keys = ['id', 'name', 'value', 'type', 'job_options']
     types = Bunch(
         REPLACEMENT_PARAMETERS='replacements',
         META_PARAMETERS='meta',  #
```
Maybe I'm missing something but there shouldn't be a new column to add - you should just add a new type here for ``RESOURCE_PARAMETERS``. Or is the simple key-value setup here too simple for what you want to do?

JeffreyThiessen replied:
You are right, I completely misunderstood how this section of code should be used.

jmchilton commented on config/workflow_resource_params_conf.xml.sample:
```diff
@@ -0,0 +1,33 @@
+			<option value="plus_ultra" label="Plus Ultra"/>
+		</param>
+	</parameters>
+	<groups default="default">
```
Using Galaxy groups this way should be possible but I would prefer it be an extension point thing and not bake into the XML format. This is how we do job parameters for instance. How about this file just defines the parameters above (so the root tag is ``root``). And then in galaxy.ini you can define a default set of parameters for all workflows -or- a Python method. That Python method will be passed all the stuff it could want - the workflow itself, the user, etc... and should return a list of parameters to display. In a second pass it could generate the parameters on the fly even.

JeffreyThiessen replied:
Could you elaborate on how/where I would define such a python method?

One of the goals we had for this change was to allow our admin to easily change which users had access to what priority levels. This is why I used the xml to define group access to parameters so we could manage the users(groups) from within galaxy and modify the xml when we need to change what groups access a specific parameter.

I agree with you that baking groups into the xml is probably not the best approach but I am not sure how I should make the changes to allow our intended functionality.

JeffreyThiessen committed:
Fixed use of WorkflowRequestInputParameter

JeffreyThiessen committed:
cleanup

JeffreyThiessen committed:
merge conflicts

JeffreyThiessen committed:
update static files after merge confict fix

JeffreyThiessen committed:
fix import order

jmchilton committed:
Refactor workflow resource parameters.

This refactoring is untested and definitely broken in a couple places - but it illustrates broad directions I'd like the work to move in.

- On the input side of things - it serializes the individual parameters to the database each as its own instance of a WorkflowRequestInputParameter. In a couple places I've retried to refactor away from the terminology of "workflow_options" toward resource parameters - there are already other kinds of workflow options.
- On the backend - I've moved a lot of the logic into galaxy.workflow.resources and tried to make it more extensible. The goal is to allow the admin to specify any kind of function to define the parameters as they want for a given workflow and user - but to implement the existing "group based" variant desired by phac-nml as just one potential variant of that with a convient YAML description.
- This moves the priority validation for workflow priorities out of the plugin and handles it up when the request is being handled. This is kind of a hack - but it results in cleaner separation between the DTD and code for defining workflow priorities so that the DTD would work with any potential workflow parameter mapping XML format or dynamic mapping Python function.
- This exposes the workflow parameters to traditional dynamic job destinations (the DTD framework re-computes it from the "job" object and that is fine).

jmchilton commented:
Thanks for the second draft - I spent most of my morning working on comments for this in the form of starting refactoring I'd like to see. I hope that is okay - I don't think my changes work at all in fact I'm fairly sure they break things that already were working - but I hope they clarify changes I would like to see.

The changes are here - https://github.com/jmchilton/galaxy/commit/22ced643434115b455473cf71eaa00b0859aca9a. I'd recommend cherry picking that into your branch and fixing it up until the code works again - but however you want to approach it is fine and I'm happy to clarify any of the changes.

- On the input side of things - it serializes the individual parameters to the database each as its own instance of a WorkflowRequestInputParameter. You definitely got closer by not defining a new table - thanks - but this goes a step forward and serializes individual parameters - it is a bit more close to what I was intended to do with that table. In a couple places I've retried to refactor away from the terminology of "workflow_options" toward resource parameters - there are already other kinds of workflow options. I hope this is okay.
- On the backend - I've moved a lot of the logic into ``galaxy.workflow.resources`` and tried to make it more extensible. The goal is to allow the admin to specify any kind of function to define the parameters as they want for a given workflow and user - but to implement the existing "group based" variant you have implemented as just one potential variant of that with a convenient YAML description. This allows swapping out the parameter specification based on pretty much anything - not just the user's Galaxy groups - you could look at LDAP, workflow tags, workflow ids, workflow names, etc... Hopefully this answers you question about where to define such a Python function.
- This moves the priority validation for workflow priorities out of the plugin and handles it up when the request is being handled. This is kind of a hack - but it results in cleaner separation between the DTD and code for defining workflow priorities so that the DTD would work with any potential workflow parameter mapping XML format or dynamic mapping Python function.
- This exposes the workflow parameters to traditional dynamic job destinations (the DTD framework re-computes it from the "job" object and that is fine).

JeffreyThiessen committed:
Added and fixed changes from 'Refactor workflow resource parameters.' by jmchilton

JeffreyThiessen committed:
merge conflicts from dev

JeffreyThiessen committed:
Remove dead code + fix merge confict mistake

JeffreyThiessen committed:
fix static files

JeffreyThiessen committed:
pep8 + tests

JeffreyThiessen committed:
undo test case change

JeffreyThiessen committed:
flake8

JeffreyThiessen committed:
fixed pep8

JeffreyThiessen committed:
fix unit test cases

JeffreyThiessen committed:
Changed 'option' to 'parameter' where appropriate

JeffreyThiessen commented:
I have gotten the code changes from https://github.com/jmchilton/galaxy/commit/22ced643434115b455473cf71eaa00b0859aca9a incorperated into this pull request and made all the changes to get it working for both yaml and python functions as `workflow_resource_params_mapper`.
I also did some more refactoring to clean up the code.

There are a few area's where some decisions need to be made for functionality and the caching function is currently not working. I have marked these area's with `TODO` and information regarding the question or problem I have.
@jmchilton It would be great if you could take a look at this in it's current state and give me some feedback on those area's.

Thanks,

JeffreyThiessen commented:
Here is the issue I am having with caching in more detail. 

Right now we have added `_resource_parameters` and the fetch/cache functions as follows.

```
class WorkflowInvocation(object, UsesCreateAndUpdateTime, Dictifiable):
    dict_collection_visible_keys = ['id', 'update_time', 'workflow_id', 'history_id', 'uuid', 'state']
    dict_element_visible_keys = ['id', 'update_time', 'workflow_id', 'history_id', 'uuid', 'state']
    states = Bunch(
        NEW='new',  # Brand new workflow invocation... maybe this should be same as READY
        READY='ready',  # Workflow ready for another iteration of scheduling.
        SCHEDULED='scheduled',  # Workflow has been scheduled.
        CANCELLED='cancelled',
        FAILED='failed',
    )

    def __init__(self):
        self.subworkflow_invocations = []
        self.step_states = []
        self.steps = []
        self._resource_parameters = None # added this one
...

...
    def fetch_resource_parameters(self):
        if self._resource_parameters is None: # Fails here, AttributeError: 'WorkflowInvocation' object has no attribute '_resource_parameters'
            self.__cache_resource_parameters()
        return self._resource_parameters

    def __cache_resource_parameters(self):
        # TODO: prefetch and optimize down the road.
        resource_type = WorkflowRequestInputParameter.types.RESOURCE_PARAMETERS
        resource_parameters = {}
        for input_parameter in self.input_parameters:
            if input_parameter.type == resource_type:
                resource_parameters[input_parameter.name] = input_parameter.value

        self._resource_parameters = resource_parameters
```

When `fetch_resource_parameters()` is called, it cannot find `_resource_parameters` on the object.

I have narrowed the issue down to `scheduling_manager.py`. When `WorkflowSchedulingManager.queue` is called it changes the `workflow_invocation` object which removes `_resource_parameters`.

```
class WorkflowSchedulingManager(object, ConfiguresHandlers):
...

...
    def queue(self, workflow_invocation, request_params):
        workflow_invocation.state = model.WorkflowInvocation.states.NEW
        scheduler = request_params.get("scheduler", None) or self.default_scheduler_id
        handler = self._get_handler(workflow_invocation.history.id)

        if handler is None and not self.__use_stack_messages:
            raise RuntimeError("Unable to set a handler for workflow invocation '%s'" % workflow_invocation.id)

        log.info("Queueing workflow invocation for handler [%s]", handler)
        workflow_invocation.scheduler = scheduler
        workflow_invocation.handler = handler

        import pdb
        pdb.set_trace() # First workflow_invocation.__dict__

        sa_session = self.app.model.context
        sa_session.add(workflow_invocation)
        sa_session.flush()

        pdb.set_trace() # Second workflow_invocation.__dict__
...
```

First `workflow_invocation.__dict__`

```
-> sa_session = self.app.model.context
(Pdb) 
(Pdb) workflow_invocation.__dict__
{'step_states': [<galaxy.model.WorkflowRequestStepState object at 0x7f74c01b4450>, <galaxy.model.WorkflowRequestStepState object at 0x7f74c01b9390>, <galaxy.model.WorkflowRequestStepState object at 0x7c0aa90>], '_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x7f74d827e390>, u'uuid': UUID('7708b03a-e436-11e7-b734-484d7ef5cf95'), '_resource_parameters': None, 'input_parameters': [<galaxy.model.WorkflowRequestInputParameter object at 0x7f74c01b41d0>, <galaxy.model.WorkflowRequestInputParameter object at 0x7f74d827eb50>, <galaxy.model.WorkflowRequestInputParameter object at 0x7f74c01b4bd0>], 'workflow': <galaxy.model.Workflow object at 0x7f74d83a2450>, u'state': 'new', u'handler': 'main', 'steps': [], u'scheduler': 'core', 'subworkflow_invocations': [], 'history': <galaxy.model.History object at 0x7f74d8262a50>}
(Pdb) 
```

Second `workflow_invocation.__dict__`

```
-> if handler is None and self.__use_stack_messages:
(Pdb) 
(Pdb) workflow_invocation.__dict__
{'_sa_instance_state': <sqlalchemy.orm.state.InstanceState object at 0x7f74d827e390>, '_resource_parameters': None}
(Pdb) 
```
If I understand how SQLAlchemy works correctly, it only adds stuff on the WorkflowInvocation table or foreign keys on it to the instance_state object. But since resource parameters are a subset of the input_parameters that already exist on the table, it doesn't make much sence to add it as another column/key.

JeffreyThiessen committed:
Merge dev into branch + merge conflict fixes

JeffreyThiessen committed:
fixed missing comma

Takadonet commented:
@jmchilton Any input to keep this pull request moving forward?

jmchilton commented:
That caching isn't needed - sorry for adding it and adding complexity. Until there is evidence there is a performance hit for not caching I'd avoid it - sorry about that.

Reviewing the rest of the TODOs today - sorry for the delay - it was the holidays and then there have been some high priority things I needed to work on for the release and for my boss.

jmchilton commented on lib/galaxy/jobs/dynamic_tool_destination.py:
```diff
@@ -1229,7 +1230,39 @@ def map_tool_to_destination(
+            # Should this really override workflow level priority? -John
+            # TODO: make a decision on what overwrites what. The reasoning behind having workflow > job is that a job
+            # parameter may exist on jobs in a workflow, and if the user runs the workflow with a parameter, they may
+            # not know that the job is overwriting their choice.
```
~It makes sense for your DTD rules that the workflow priority would take precedents for sure. I just feel like this file treats the parameters without any semantic meaning - they are just keys and values. So it would be better to structure this in such a way that the dynamic tool destinations plugin made that determination. I'll keep looking and see if I see an obvious way to do that.~

This is in DTD. I think I thought this piece of code was the job priority override the workflow priority - if it is the workflow priority override the job priority then great - delete my comment I think.

jmchilton commented on lib/galaxy/model/__init__.py:
```diff
@@ -4146,6 +4148,37 @@ def add_input(self, content, step_id):
+    #
+    #     self._resource_parameters = resource_parameters
+
+    # TODO: Remove this. Temp solution until caching is fixed.
```
This is good for now. Drop the TODO and delete the cached stuff I think.

jmchilton replied:
Add a ``@property`` decorator to this method and drop the parens in the next TODO down. https://docs.python.org/2/library/functions.html#property

jmchilton commented on lib/galaxy/workflow/resources/example.py:
```diff
@@ -0,0 +1,22 @@
+import logging
```
Can you ``git rm`` this file from the git index - I guess only the sample should be in here.

jmchilton commented on lib/galaxy/workflow/resources/__init__.py:
```diff
@@ -0,0 +1,176 @@
+    # workflow_resource_params so that we disallow selection when the yml is not correct since that file would be
+    # changed the most frequently and most prone to configuration errors.
+
+    # TODO: finish validation? Currently validating needed stuff except for 'options' in a 'select'
```
This validation is fine - it only applies to the by_group variant of this - which I think is what I was worried about. If ultimately "by_group" could look a bit more like a plugin and not be dispatched in ifs so much - that would be more esthetically pleasing but it isn't actually important I don't think. Clear out the TODOs here if it is working for you?

jmchilton commented on lib/galaxy/util/__init__.py:
```diff
@@ -817,6 +817,22 @@ def xml_text(root, name=None):
+def parse_resource_parameters(resource_param_file):
+    """Code shared between jobs and workflows for reading resource parameter configuration files.
+
+    TODO: Allow YAML in addition to XML.
```
This is my TODO - it can just be left in the code.

jmchilton commented:
I'm thinking this looks really nice @JeffreyThiessen - thanks so much for this and sorry again for not reviewing it in a timely manner.

This also needs a merge against dev I think - which is going to be tricky because the galaxy.ini.sample file is gone. I think you will want to place those option documentation changes in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/config_schema.yml land then regenerate the sample YAML file (that has replaced the sample ini file) using ``config-rebuild-sample``. Also we aren't tracking static file changes in dev anymore - so it'd be best to remove the changes to the map files and such modified in this PR. Ideally this could all be done with a rebase to clean up the history and such - but a merge can work also I think. If you want me to do a rebase and push an updated git history to your branch let me know.

Thanks again and sorry again.

JeffreyThiessen commented:
Thanks for the feedback, I'll get those changes done as soon as possible.

JeffreyThiessen committed:
remove/fix todo's

JeffreyThiessen committed:
merge conflicts

JeffreyThiessen committed:
move vars from galaxy.ini to galaxy.yml

JeffreyThiessen committed:
merge in dev

JeffreyThiessen committed:
merge dev into branch

JeffreyThiessen committed:
fix by group validation

JeffreyThiessen committed:
remove static files

JeffreyThiessen committed:
merge from dev

JeffreyThiessen commented:
@jmchilton This PR should be ready to merge now. I added options validation to the by_group validation so the whole mapper file will be validated now.

~~I have only been able to test it by running galaxy with the galaxy.ini from 17.09 though. With the galaxy.yml file I cannot get any jobs to run with the DTD, with or without this PR change so I think that problem is unrelated.~~ Problem resolved, was due to misconfigured `job_conf.xml`

JeffreyThiessen committed:
fix merge conflict

jmchilton commented:
@galaxybot test this

jmchilton commented:
Thanks for the merge and such, I've removed the WIP tag. It looks ready to go to me - I want to see what the tests say first though - don't worry if a few are red - I'll let you know if you broke anything. Thanks for your patience.

JeffreyThiessen commented:
No problem, @Takadonet says I can pass my last Co-op term now :laughing:
